File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/data/generators/time-parameters.ts
import { TimeGranularity } from '../../time/time-granularity.enum';
import { BusinessDayCalendar } from '../../time/business-day-calendar.enum';
import {
  objectName, sessionStartTimeName, sessionEndTimeName, startDateName, timeGranularityName,
  businessDayCalendarName
} from './constants';

/** The time-related input parameters for temporal data generators. */
export class TimeParameters {
  private static readonly defaultSessionStartTime: string = '09:00:00';
  private static readonly defaultSessionEndTime: string = '18:00:00';
  private static readonly dafaultStartDate: Date = new Date(2000, 0, 1);
  private static readonly dafaultTimeGranularity: TimeGranularity = TimeGranularity.Day1;
  private static readonly dafaultBusinessDayCalendar: BusinessDayCalendar = BusinessDayCalendar.WeekendsOnly;

  /** The start time of the trading session. */
  sessionStartTime: string = TimeParameters.defaultSessionStartTime;

  /** The end time of the trading session. */
  sessionEndTime: string = TimeParameters.defaultSessionEndTime;

  /** The date of the first data sample. */
  startDate: Date = TimeParameters.dafaultStartDate;

  /** The time granularity of data samples. */
  timeGranularity: TimeGranularity = TimeParameters.dafaultTimeGranularity;

  /** An exchange holiday schedule or a general country holiday schedule. */
  businessDayCalendar: BusinessDayCalendar = TimeParameters.dafaultBusinessDayCalendar;

  constructor(data?: TimeParameters) {
    if (data) {
      for (const property in data) {
        if (data.hasOwnProperty(property)) {
          (this as any)[property] = (data as any)[property];
        }
      }
    }
  }

  private static formatDate(d: Date): string {
    return d.getFullYear() + '-' +
      (d.getMonth() < 9 ? ('0' + (d.getMonth() + 1)) : (d.getMonth() + 1)) + '-' +
      (d.getDate() < 10 ? ('0' + d.getDate()) : d.getDate());
  }

  toJSON(data?: any) {
    data = typeof data === objectName ? data : {};
    data[sessionStartTimeName] = this.sessionStartTime;
    data[sessionEndTimeName] = this.sessionEndTime;
    data[startDateName] = this.startDate ? TimeParameters.formatDate(this.startDate) : TimeParameters.dafaultStartDate;
    data[timeGranularityName] = this.timeGranularity;
    data[businessDayCalendarName] = this.businessDayCalendar;
    return data;
  }
}

File: Mbs/Trading/Data/Generators/TimeParameters.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using Mbs.Trading.Time;

namespace Mbs.Trading.Data.Generators
{
    /// <summary>
    /// The time-related input parameters for temporal data generators.
    /// </summary>
    public class TimeParameters : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the start time of the trading session.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DataType(DataType.Time)]
        [DefaultValue(DefaultParameterValues.SessionStartTimeString)]
        public TimeSpan SessionStartTime { get; set; } = DefaultParameterValues.SessionStartTime;

        /// <summary>
        /// Gets or sets the end time of the trading session.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DataType(DataType.Time)]
        [DefaultValue(DefaultParameterValues.SessionEndTimeString)]
        public TimeSpan SessionEndTime { get; set; } = DefaultParameterValues.SessionEndTime;

        /// <summary>
        /// Gets or sets the date of the first data sample.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DataType(DataType.Date)]
        [DefaultValue(DefaultParameterValues.StartDateString)]
        public DateTime StartDate { get; set; } = DefaultParameterValues.StartDate;

        /// <summary>
        /// Gets or sets the time granularity of data samples.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [EnumDataType(typeof(TimeGranularity), ErrorMessage = ErrorMessages.FieldEnumValueInvalid)]
        [DefaultValue(DefaultParameterValues.TimeGranularity)]
        public TimeGranularity TimeGranularity { get; set; } = DefaultParameterValues.TimeGranularity;

        /// <summary>
        /// Gets or sets a value specifying an exchange holiday schedule or a general country holiday schedule.
        /// Business days do not form a continuous sequence (weekends, holidays etc.),
        /// so there is a need in differentiating between the business time and the physical time.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [EnumDataType(typeof(BusinessDayCalendar), ErrorMessage = ErrorMessages.FieldEnumValueInvalid)]
        [DefaultValue(DefaultParameterValues.BusinessDayCalendar)]
        public BusinessDayCalendar BusinessDayCalendar { get; set; } = DefaultParameterValues.BusinessDayCalendar;

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            if (SessionStartTime >= SessionEndTime)
            {
                yield return new ValidationResult(
                    $"{nameof(SessionStartTime)} {SessionStartTime} should be less than the {nameof(SessionEndTime)} {SessionEndTime}.",
                    new[] { nameof(SessionStartTime), nameof(SessionEndTime) });
            }
        }
    }
}
